File: src/components/message_list/states.py
from utils.toolkit import Gtk, Adw
from components.ui import AppIcon, AppText

class MessageListStates:
    def __init__(self, widget):
        self.widget = widget
        self.empty_state = self._create_empty_state()
        self.loading_state = self._create_loading_state()
        self.error_state = self._create_error_state()
        self.error_text = None

    def _create_empty_state(self):
        container = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        container.add_css_class("message-list-empty-state")
        container.set_halign(Gtk.Align.CENTER)
        container.set_valign(Gtk.Align.CENTER)
        container.set_spacing(12)
        container.set_vexpand(True)  
        container.set_hexpand(True)

        icon = AppIcon("mail-unread-symbolic", class_names="message-list-empty-icon")
        icon.set_pixel_size(48)
        icon.set_opacity(0.5)

        text = AppText(
            "No messages in this folder",
            halign=Gtk.Align.CENTER,
            class_names="message-list-empty-text",
        )
        text.set_opacity(0.7)

        container.append(icon.widget)
        container.append(text.widget)

        return container

    def _create_loading_state(self):
        container = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        container.add_css_class("message-list-loading-state")
        container.set_halign(Gtk.Align.CENTER)
        container.set_valign(Gtk.Align.CENTER)
        container.set_spacing(12)

        spinner = Gtk.Spinner()
        spinner.add_css_class("message-list-loading-spinner")
        spinner.set_spinning(True)

        text = AppText(
            "Loading messages...",
            halign=Gtk.Align.CENTER,
            class_names="message-list-loading-text",
        )

        container.append(spinner)
        container.append(text.widget)

        return container

    def _create_error_state(self):
        container = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        container.add_css_class("message-list-error-state")
        container.set_halign(Gtk.Align.CENTER)
        container.set_valign(Gtk.Align.CENTER)
        container.set_spacing(12)

        icon = AppIcon("dialog-error-symbolic", class_names="message-list-error-icon")
        icon.set_pixel_size(48)
        icon.set_opacity(0.5)

        self.error_text = AppText(
            "Failed to load messages",
            halign=Gtk.Align.CENTER,
            class_names="message-list-error-text",
        )
        self.error_text.set_opacity(0.7)

        container.append(icon.widget)
        container.append(self.error_text.widget)

        return container

    def show_empty(self):
        self.hide_all()
        self.widget.add(self.empty_state)

    def show_loading(self):
        self.hide_all()
        self.widget.add(self.loading_state)

    def show_error(self, error_message=None):
        self.hide_all()
        self.widget.add(self.error_state)
        if error_message and self.error_text:
            self.error_text.set_text_content(f"Failed to load messages: {error_message}")

    def show_list(self):
        self.hide_all()

    def hide_all(self):
        if self.empty_state.get_parent():
            self.widget.remove(self.empty_state)
        if self.loading_state.get_parent():
            self.widget.remove(self.loading_state)
        if self.error_state.get_parent():
            self.widget.remove(self.error_state) 

File: src/components/states.rs
use super::{column, label};
use crate::theme;
use adw::prelude::*;

pub fn placeholder(text: &str, class: &str, icon: &str, loading: bool) -> gtk::Box {
    let content = column(class);
    content.set_halign(gtk::Align::Center);
    content.set_valign(gtk::Align::Center);
    content.set_hexpand(true);
    content.set_vexpand(true);
    content.set_spacing(theme::SPACING);
    if loading {
        let spinner = gtk::Spinner::builder()
            .spinning(true)
            .width_request(theme::AVATAR_SIZE)
            .height_request(theme::AVATAR_SIZE)
            .build();
        content.append(&spinner);
    } else {
        let image = gtk::Image::from_icon_name(icon);
        image.set_pixel_size(theme::STATE_ICON);
        image.set_opacity(0.5);
        content.append(&image);
    }
    let text = label(text, "state-text");
    text.set_halign(gtk::Align::Center);
    text.set_opacity(0.7);
    text.set_wrap(true);
    content.append(&text);
    content
}

pub fn select_message(viewer: &gtk::Box) {
    super::clear(viewer);
    viewer.set_vexpand(true);
    viewer.append(&placeholder(
        "Select message to view",
        "message-viewer-empty-state",
        "mail-unread-symbolic",
        false,
    ));
}

pub fn list_state(stack: &gtk::Stack, text: &str, loading: bool, error: bool) {
    if let Some(previous) = stack.child_by_name("state") {
        stack.remove(&previous);
    }
    let class = if error {
        "message-list-error-state"
    } else if loading {
        "message-list-loading-state"
    } else {
        "message-list-empty-state"
    };
    let state = placeholder(
        text,
        class,
        if error {
            "dialog-error-symbolic"
        } else {
            "mail-unread-symbolic"
        },
        loading,
    );
    if error {
        state.add_css_class("error");
    }
    stack.add_named(&state, Some("state"));
    stack.set_visible_child_name("state");
}

pub fn refreshing(button: &gtk::Button, loading: bool) {
    if loading {
        button.set_child(Some(&gtk::Spinner::builder().spinning(true).build()));
    } else {
        button.set_icon_name("view-refresh-symbolic");
    }
}
